### PR TITLE
[QeRL] Compose online quantization with quantized reloading

### PR DESCRIPTION
**Pull request by kylesayrs**
## Purpose ##
* Support online quantized reloading
* Reuse layerwise reloading logic for online quantization
  * Only convert fp8 for now, support mxfp8 in a follow-up

## Lifecycle ##
Online quantization follows a similar lifecycle to reloading:

| Step | Quantized Reload | Online Quantization |
| - | - | - |
| `record_metadata_for_reloading` | Record tensor metadata so that layers can be restored on the meta device | Called but not used |
| `restore_layer_on_meta` | Restore layer to model format at start of reload | Not called |
| `initialize_online_processing` | Wrap weight loaders with `online_process_loader`, which buffers weights until all layer weights are ready | Called by layers with online quantization |
| `_layerwise_process` | Process weights once all weights are loaded | Called by layers with online quantization to quantize weights |
| Copy into kernel tensors | Copy processed weights into original tensor locations to affect compiled cuda graphs, ect. | Skipped |
| `finalize_layerwise_reload` | Catch any layers which did not load all weights (due to attention or padding) | Called after loading, supports dummy format |

## Changes ##
* Break out `initialize_online_processing` for use by online quantization methods
  * Online quantization methods instantiate the full precision weight on the meta device, then call `initialize_online_processing` on the layer to add layerwise wrappers
  * By reusing the layerwise `online_process_loader`, we can ensure that weight loaders are never wrapped twice during online quantized reloading
* Skip kernel copying and replacement step when online quantizing (not reloading)
* Call `finalize_layerwise_process` in `BaseModelLoader.load_weights`. In the future, this function could fully replace `process_weights_after_loading`

## Edge case handling ##
* Some models have loaded parameters which are not created by the quantization method, for example `e_score_correction_bias`. Ideally, these parameters would be initialized on the meta device just like the weights. However, quant_method has no control over these parameters, so we end up loading them twice, leading to a small performance loss

## Testing ##
* Regression tested `test_online_quantization` and `test_reload_weights`
* Test online quantized reloading in `test_online_quantize_reload`

### CODE REVIEW

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 1]
```diff
@@ -185,16 +192,18 @@ def finalize_layerwise_reload(model: torch.nn.Module, model_config: ModelConfig)
                 )
 
             else:
-                _place_kernel_tensors(layer, info)
+                # reloading: place kernel tensors back
+                if info.kernel_tensors is not None:
+                    _place_kernel_tensors(layer, info)
                 layer.process_weights_after_loading(model_config.dtype)
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py:208]: Attention is weird, `process_weights_after_loading` is defined on the module

File: vllm/model_executor/model_loader/reload/meta.py [hunk 2]
```diff
@@ -122,7 +122,7 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         if kwargs is None:
             kwargs = {}
 
-        if func is torch.ops.aten.copy_.default and args[0].device.type == "meta":
+        if func is torch.ops.aten.copy_.default:
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/meta.py:125]: This should be safe, and is unlikely to have bad/ unnoticed outcomes. This is necessary because some weights are initialized outside of `create_weights`, and therefore we cannot force them to be initialized on the meta device.

File: vllm/model_executor/layers/quantization/fp8.py [hunk 3]
```diff
@@ -1105,91 +964,47 @@ def patched_weight_loader(param, loaded_weight, *args, **kwargs):
...
         # BIASES (for models like GPT-OSS that have biased MoE)
         if self.moe.has_bias:
-            # Use the original weight_loader (not patched) for biases
-            orig_extra_weight_attrs = dict(extra_weight_attrs)
-            orig_extra_weight_attrs["weight_loader"] = weight_loader
             w13_bias = torch.nn.Parameter(
                 torch.zeros(
                     num_experts,
                     2 * intermediate_size_per_partition,
+                    device="meta",
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/fp8.py]: can we tell the user where this is intialized

File: tests/model_executor/model_loader/test_reload.py [hunk 4]
```diff
@@ -148,3 +148,60 @@ def test_reload_weights(base_model, mul_model, add_model, tp_size, vllm_runner):
...
+    if quantization == "fp8" and not current_platform.supports_fp8():
+        pytest.skip(reason="Requires FP8 support")
+
+    with vllm_runner(
+        model_name=base_model,
+        quantization=quantization,
+        tensor_parallel_size=tp_size,
+        enable_expert_parallel=(tp_size > 1 and "DeepSeek" in base_model),
+        enable_prefix_caching=False,
+    ) as llm:
```
Comment (vkuzo) [inline tests/model_executor/model_loader/test_reload.py:198]: can we test behavior after the model is loaded and before the first reload happens

File: vllm/model_executor/layers/quantization/fp8.py [hunk 5]
```diff
@@ -514,84 +518,22 @@ def create_weights(
...
-
-            return res
-
         weight = ModelWeightParameter(
             data=torch.empty(
                 output_size_per_partition,
                 input_size_per_partition,
-                # materialized just-in-time in `patched_weight_loader`
                 device="meta",
                 dtype=params_dtype,
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/fp8.py:568]: it would be good to keep this comment and just update for the new location, the code reader should know where to go if they want to understand how the materialization happens

File: vllm/model_executor/layers/quantization/fp8.py [hunk 6]
```diff
@@ -1012,86 +945,12 @@ def create_weights(
...
-
         # WEIGHTS
         w13_weight = torch.nn.Parameter(
             torch.empty(
                 num_experts,
                 2 * intermediate_size_per_partition,
                 hidden_size,
-                # materialized just-in-time in `patched_weight_loader`
                 device="meta",
                 dtype=params_dtype,
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/fp8.py:1096]: nit: keep a comment

File: vllm/model_executor/layers/quantization/fp8.py [hunk 7]
```diff
@@ -1105,91 +964,47 @@ def patched_weight_loader(param, loaded_weight, *args, **kwargs):
                 num_experts,
                 hidden_size,
                 intermediate_size_per_partition,
-                # materialized just-in-time in `patched_weight_loader`
                 device="meta",
                 dtype=params_dtype,
```
Comment (vkuzo) [inline vllm/model_executor/layers/quantization/fp8.py:1110]: nit: keep a comment

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 8]
```diff
@@ -239,18 +255,31 @@ def _layerwise_process(layer: torch.nn.Module, info: LayerReloadingInfo):
...
+            buffer.data.copy_(getattr(layer, name))
 
-    _place_kernel_tensors(layer, info)
+        _place_kernel_tensors(layer, info)
 
     info.reset()
     logger.debug("%s: Processed", layer.__class__.__name__)
 
 
+def _place_or_materialize(layer: torch.nn.Module, info: LayerReloadingInfo):
```
Comment (vkuzo) [inline vllm/model_executor/model_loader/reload/layerwise.py]: this is a bit confusing, can we just branch at the callsite instead of here?

File: vllm/model_executor/model_loader/reload/types.py [hunk 9]
```diff
@@ -17,7 +17,7 @@ class LayerReloadingInfo:
     restore_metadata: LayerTensors = field(default_factory=lambda: ({}, {}))
 
     # kernel format (device)
-    kernel_tensors: LayerTensors = field(default_factory=lambda: ({}, {}))
+    kernel_tensors: LayerTensors | None = None
```
Comment (vkuzo) [inline vllm/model_executor/model_loader/reload/types.py:20]: nit: can we make the comment explain when this is none vs specified

**Review by vkuzo — APPROVED** [review]
thank you! my comments are optional nits, lg if tests pass

File: tests/model_executor/model_loader/test_reload.py (hunk 4 above)
Comment (kylesayrs) [reply tests/model_executor/model_loader/test_reload.py:198]: One of the tested models is empty `inference-optimization/DeepSeek-V3-debug-empty`, so won't produce sane results.

Rather than add a kluge, I'd rather rely on specific online quantization tests, ie `tests/quantization/test_fp8.py::test_online_quantization`. Do you think this is enough coverage?

File: vllm/model_executor/model_loader/base_loader.py [hunk 10]
```diff
@@ -84,3 +87,12 @@ def log_model_inspection(model: nn.Module) -> None:
     from vllm.model_inspection import format_model_inspection
 
     logger.info("vLLM model structure:\n%s", format_model_inspection(model))
+
+
+def _has_online_quant(model: nn.Module):
```
Comment (kylesayrs) [inline vllm/model_executor/model_loader/base_loader.py:92]: This kluge will be removed by https://github.com/neuralmagic/vllm/pull/153

File: tests/model_executor/model_loader/test_reload.py (hunk 4 above)
Comment (vkuzo) [reply tests/model_executor/model_loader/test_reload.py:198]: could just skip this for for `DeepSeek-V3-debug-empty`?

I do think we should test for sane output after online quant and before the first reloading call, `test_fp8.py` does not test anything related to layerwise reloading

**Review by mgoin — APPROVED** [review]
Changes look nice to me, and I'm fairly certain this can't affect the non-online path so I feel good about the spurious failures.

File: vllm/model_executor/model_loader/base_loader.py [hunk 11]
```diff
@@ -49,16 +50,13 @@ def load_model(
             device_config.device if load_config.device is None else load_config.device
         )
         target_device = torch.device(load_device)
-        with set_default_torch_dtype(model_config.dtype):
-            with target_device:
-                model = initialize_model(
-                    vllm_config=vllm_config, model_config=model_config, prefix=prefix
-                )
-
+        with set_default_torch_dtype(model_config.dtype), target_device:
```
Comment (jikunshang) [inline vllm/model_executor/model_loader/base_loader.py:53]: I feel this may not safe. and I guess it cause `Language Models Tests (Extra Standard) 2` case fail.
see #38426

Comment (kylesayrs) [reply vllm/model_executor/model_loader/base_loader.py:53]: @jikunshang I agree, although I'm not 100% sure why. I'm still trying to figure out why/ how to have both
